channeldb/kvdb/etcd/stm.go

```diff
@@ -227,39 +236,61 @@ func runSTM(s *stm, apply func(STM) error) error {
 		err     error
 	)
 
-loop:
-	// In a loop try to apply and commit and roll back if the database has
-	// changed (CommitError).
-	for {
-		select {
-		// Check if the STM is aborted and break the retry loop if it is.
-		case <-s.options.ctx.Done():
-			err = fmt.Errorf("aborted")
-			break loop
+	done := make(chan struct{})
 
-		default:
-		}
-		// Apply the transaction closure and abort the STM if there was
-		// an application error.
-		if err = apply(s); err != nil {
-			break loop
-		}
+	execute := func() {
+		defer close(done)
+
+		for {
+			select {
+			// Check if the STM is aborted and break the retry loop
+			// if it is.
+			case <-s.options.ctx.Done():
+				err = fmt.Errorf("aborted")
+				return
+
+			default:
+			}
 
-		stats, err = s.commit()
+			stats, err = s.commit()
 
-		// Retry the apply closure only upon commit error (meaning the
-		// database was changed).
-		if _, ok := err.(CommitError); !ok {
-			// Anything that's not a CommitError aborts the STM
-			// run loop.
-			break loop
+			// Re-apply only upon commit error (meaning the
+			// keys were changed).
+			if _, ok := err.(CommitError); !ok {
+				// Anything that's not a CommitError
+				// aborts the transaction.
+				return
+			}
+
+			// Rollback before trying to re-apply.
+			s.Rollback()
+			retries++
+
+			// Re-apply the transaction closure.
+			if err = apply(s); err != nil {
+				return
+			}
 		}
+	}
 
-		// Rollback before trying to re-apply.
-		s.Rollback()
-		retries++
+	// Run the tx closure to construct the read and write sets.
+	// Also we expect that if there are no conflicting transactions
+	// in the queue, then we only run apply once.
+	if err = apply(s); err != nil {
```

Comment:
Yes, so it's a bit tricky to read this at first, but it's really simple actually.
1) What we do is we first run the `apply` closure to gather the read/write sets so we can add the tx to the contention queue.
2) The `execute` closure is executed there (either immediately or in the queue goroutine).

The `err` simply holds the error trough the above described execution graph.
3) we wait for the `done` signal and them clean the keys from the queue.

Reply:
thanks for the explanation, makes sense now! i also see that it's the return value at the end of the function, so that's where it is "read"

Reply:
Since we return immediately if the error is non-nil here, wouldn't `if err := apply(s); err != nil ` be equivalent? 

I agree that the shadowing is tricky to read. Could make sense to add more errors with descriptive names (i.e. `executeErr`) to make it easier.

Reply:
Yeah, maybe it's simpler to read if we distinguish errors by scope. PTAL